mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulPlayScreen.kt

```diff
@@ -1,34 +1,189 @@
+ */
+private inline fun <A> Loadable<A>.orElse(lazyBlock: () -> A): A =
+    when (this) {
+      is Loadable.Loaded -> value
+      Loadable.Loading -> lazyBlock()
+    }
+
+/**
+ * Gives higher order function to decide the MatchResult
+ * @param userColor color of the current user
+ */
+private inline fun NextStep.toMatchResult(): (userColor: Color) -> MatchResult =
+    when (this) {
+      is NextStep.Stalemate -> { _ -> Tie }
+      is NextStep.MovePiece -> { userColor -> if (this.turn == userColor) YourTurn else OtherTurn }
+      is NextStep.Checkmate -> { userColor ->
+            if (userColor == this.winner) {
+              Win(MatchResult.Reason.CHECKMATE)
+            } else {
+              Loss(MatchResult.Reason.CHECKMATE)
+            }
+          }
+    }
+
+/** Intermediate State to simplify the data handling */
+private data class MatchInfo(
+    val id: String?,
+    val movesCount: Int,
+    val whiteId: String,
+    val blackId: String,
+    val whiteName: String,
+    val blackName: String,
+    val matchResult: (userColor: Color) -> MatchResult
+)
+
+/**
+ * Fetches the matches from the facade and convert it to a list of MatchInfos
+ * @param currentUser the current authenticated user
+ * @param facade the faced we fetch the matches
+ */
+private fun fetchForUser(
+    currentUser: AuthenticatedUser,
+    facade: ChessFacade,
+): Flow<List<MatchInfo>> =
+    facade.matches(currentUser).map { m -> m.map { info(it) } }.flatMapLatest { matches ->
+      combine(matches) { it.toList() }
+    }
+
+/**
+ * Extract the data from the subflows of black and white and combine all the information in
+ * MatchInfo
+ * @param match a [Match] to extract the informations
+ */
+private fun info(match: Match): Flow<MatchInfo> {
+  val black = match.black
+  val white = match.white
+  val game = match.game
+
+  val blackName = black.map { loaded(it) }.onStart { emit(loading()) }
+  val whiteName = white.map { loaded(it) }.onStart { emit(loading()) }
+  val loadingGame = game.map { loaded(it) }.onStart { emit(loading()) }
+
+  return combine(blackName, whiteName, loadingGame) { b, w, g ->
+    val blackId = b.map { it?.uid ?: "" }.orElse { "" }
+    val whiteId = w.map { it?.uid ?: "" }.orElse { "" }
+    MatchInfo(
+        id = match.id,
+        movesCount = g.map { it.serialize().size }.orElse { 0 },
+        whiteId = whiteId,
+        blackId = blackId,
+        whiteName = w.map { it?.name ?: "" }.orElse { "" },
+        blackName = b.map { it?.name ?: "" }.orElse { "" },
+        matchResult = g.map { it.nextStep.toMatchResult() }.orElse { { Tie } })
+  }
+}
+
+/**
+ * Implementation of the PlayscreenState
+ * @param onNewGameClick callback function for the new game button
+ * @param user authenticated user
+ * @param facade [ChessFacade] to fetch the matches
+ * @param scope for coroutines
+ */
+private class PlayScreenStateImpl(
+    onNewGameClick: State<() -> Unit>,
+    override val onGameItemClick: (ChessMatch) -> Unit,
```

Comment:
Like `onNewGameClick`, you should pass this as a `State` and use a `rememberUpdatedState` when creating the `PlayScreenStateImpl`